op_builder/async_io.py

```diff
@@ -59,6 +90,15 @@ def is_compatible(self):
         aio_compatible = self.has_function('io_submit', ('aio', ))
```

Comment:
I get a compile error when I build DeepSpeed from source using install.sh. I don't fully understand what this function is doing, but it generates and fails to compile the following source.
```c
main (int argc, char **argv) {
    io_submit();
}
```
The error message is reproducible by running gcc directly on the file:
```bash
/tmp/io_submitikjqrmn5.c:1:1: warning: return type defaults to ‘int’ [-Wimplicit-int]
 main (int argc, char **argv) {
 ^~~~
/tmp/io_submitikjqrmn5.c: In function ‘main’:
/tmp/io_submitikjqrmn5.c:2:5: warning: implicit declaration of function ‘io_submit’ [-Wimplicit-function-declaration]
     io_submit();
     ^~~~~~~~~
/tmp/ccFfhIAq.o: In function `main':
io_submitikjqrmn5.c:(.text+0x15): undefined reference to `io_submit'
collect2: error: ld returned 1 exit status
```

So  I am curious how this is working for you (and in our CI). Can you help? Thanks.

Reply:
I accidentally hit approve, sorry for the confusion

Reply:
The ``has_function`` method is creating a simple C program that uses a named function, which it then attempts to compile and link.  In this case, the link is failing with ``undefined reference to `io_submit'``.  That function should be provided in ``libaio.so``, so one problem might be that's failing to link with libaio.

Does this cause your whole build fail, or just the test to detect whether libaio is available?

I'll see if I can reproduce it.  Can you share the ``install.sh`` command you're using?

Reply:
Yes, it is failing the entire build. Sorry for not linking [install.sh](https://github.com/microsoft/DeepSpeed/blob/master/install.sh) earlier. It is the standard script for building from source.

Thanks for the explanation. As far as I can see that c program is not well-formed since it is missing the `#include<libaio.h>` at the minimum. Also, `io_submit` takes arguments so I don't expect this call `io_submit()` to succeed. Thanks.

Reply:
Yes, that C program is not exactly well-formed, but something like that is typically enough to test whether a certain function symbol exists.  There are two problems I'd like to figure out.

First, I can imagine one gets this error if the test program fails to link with libaio or the libaio that it finds does not provide the ``io_submit`` symbol for some reason.

Second, I'd like to figure out why the failure of that test link kills the whole build rather than just disables the libaio extension.

I can force a similar type of error message if I replace ``io_submit`` with some bogus function name that I know does not exist.  However, in my case, the build seems to just disable the aio op and keep going rather than aborting the whole build.  I'm running a ``python setup.py build_dist`` command directly rather than using the ``install.sh`` script.  I still need to try that myself.

As for that second item where it kills your build, do you see any other errors messages in the ``install.sh`` build output?

Reply:
add `try/except`?

Reply:
as I mentioned earlier, this program may crash as well if the user doesn't cc tools installed.

Reply:
The ``ccompiler.has_function`` method doesn't seem to raise an exception here, even when the link fails due to an undefined symbol.  It just seems to return ``False`` in that case.

I'm reworking this to use ``compile`` and ``link_executable`` instead, which do seem to raise exceptions.  That should also let us catch compile errors (e.g., no compiler) separately from link errors (``io_submit`` is undefined).  That requires one to put together the test program manually.

I think that will help resolve the missing compiler problem.  However, I'm unsure whether this will help with either of the other two problems above, namely, why is ``io_submit`` missing, and why is that causing the whole build to abort rather than just disabling the aio op.  We'll keep hunting those.

Reply:
You're trying to solve a very complex problem, Adam. This is super challenging, and we haven't even tested it yet on a bunch of different platforms. I'm very grateful you were inspired to work on this.

If you find a way for this compile test to gracefully fail, I think it'd be a huge help - that's to have it optional and not getting in the way of the build.

Reply:
> Yes, it is failing the entire build. Sorry for not linking [install.sh](https://github.com/microsoft/DeepSpeed/blob/master/install.sh) earlier. It is the standard script for building from source.
> 
> Thanks for the explanation. As far as I can see that c program is not well-formed since it is missing the `#include<libaio.h>` at the minimum. Also, `io_submit` takes arguments so I don't expect this call `io_submit()` to succeed. Thanks.

Apologies for the delayed response. Below is the error I see with master (without your latest changes).  I will share the results of your branch asap. Also, to echo @stas00, thanks for doggedly taking on a very challenging problem.

```bash
worker-0: Installing collected packages: deepspeed
worker-0: Successfully installed deepspeed-0.4.4+6ae756c
worker-0: Warning: Permanently added '[192.168.0.52]:44896' (ECDSA) to the list of known hosts.
worker-0: /tmp/io_submitlnvp9xtf.c:1:1: warning: return type defaults to 'int' [-Wimplicit-int]
worker-0:  main (int argc, char **argv) {
worker-0:  ^~~~
worker-0: /tmp/io_submitlnvp9xtf.c: In function 'main':
worker-0: /tmp/io_submitlnvp9xtf.c:2:5: warning: implicit declaration of function 'io_submit' [-Wimplicit-function-declaration]
worker-0:      io_submit();
worker-0:      ^~~~~~~~~
worker-0: --------------------------------------------------
worker-0: DeepSpeed C++/CUDA extension op report
worker-0: --------------------------------------------------
worker-0: NOTE: Ops not installed will be just-in-time (JIT) compiled at
worker-0:       runtime if needed. Op compatibility means that your system
worker-0:       meet the required dependencies to JIT install the op.
worker-0: --------------------------------------------------
worker-0: JIT compiled ops requires ninja
worker-0: ninja .................. [OKAY]
worker-0: --------------------------------------------------
worker-0: op name ................ installed .. compatible
worker-0: --------------------------------------------------
worker-0: cpu_adam ............... [NO] ....... [OKAY]
worker-0: fused_adam ............. [NO] ....... [OKAY]
worker-0: fused_lamb ............. [NO] ....... [OKAY]
worker-0: sparse_attn ............ [NO] ....... [OKAY]
worker-0: transformer ............ [NO] ....... [OKAY]
worker-0: stochastic_transformer . [NO] ....... [OKAY]
worker-0: async_io ............... [NO] ....... [OKAY]
worker-0: transformer_inference .. [NO] ....... [OKAY]
worker-0: utils .................. [NO] ....... [OKAY]
worker-0: quantizer .............. [NO] ....... [OKAY]
worker-0: --------------------------------------------------
worker-0: DeepSpeed general environment info:
worker-0: torch install path ............... ['/opt/conda/lib/python3.6/site-packages/torch']
worker-0: torch version .................... 1.8.0a0+17f8c32
worker-0: torch cuda version ............... 11.1
worker-0: nvcc version ..................... 11.1
worker-0: deepspeed install path ........... ['/opt/conda/lib/python3.6/site-packages/deepspeed']
worker-0: deepspeed info ................... 0.4.4+6ae756c, 6ae756c, master
worker-0: deepspeed wheel compiled w. ...... torch 1.8, cuda 11.1
worker-0: Warning: Permanently added '[192.168.0.52]:44896' (ECDSA) to the list of known hosts.
```

Reply:
Thanks @tjruwase .  In the output about, it looks like it's printing the compiler warnings about the implicit return types, but I don't see the linker error about ``io_submit`` being undefined.  Actually, the bottom part of the output suggests that it left the async_io op enabled, which probably means it must have also linked successfully.

Do you also see the linker error in your ``install.sh`` output?

I've seen people suggest ways to silence the compile/link output from distutils to hide those warnings by redirecting stderr with ``os.dup2``.  For now, I'm not hiding that output since it should help us debug the problem.